### PR TITLE
Fix and small improvement to the detection algorithm

### PR DESCRIPTION
**Pull request by loikki**
Blue corner position is now in system parameter

### CODE REVIEW

File: sbc_compassion/models/import_letters_history.py [hunk 1]
```diff
@@ -246,6 +246,7 @@ def _analyze_attachment(self, file_, filename):
         :param str file_: Path of the file to analyze
         :param str filename: Filename to give in odoo
         """
+        print file_
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: Remove print

File: sbc_compassion/models/import_letters_history.py [hunk 2]
```diff
@@ -271,9 +272,12 @@ def _analyze_attachment(self, file_, filename):
...
         # loop over all the patterns in the pattern directory
         template, pattern_center = self._find_template(img)
-        lang_id = self._find_language(img, pattern_center, template)
+        if template != self.env.ref('sbc_compassion.default_template'):
+            lang_id = self._find_language(img, pattern_center, template)
+        else:
+            lang_id = None
 
-        if self.force_template:
+        if self.force_template and self.force_template is not None:
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: Is this really needed ? If `force_template` is None it evaluates to False...

File: sbc_compassion/models/import_letters_history.py [hunk 3]
```diff
@@ -426,7 +430,13 @@ def _find_template(self, img):
         """
         templates = self.env['sponsorship.correspondence.template'].search(
             [('pattern_image', '!=', False)])
-        return pr.find_template(img, templates)
+        return_value = pr.find_template(img, templates)
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: `find_template` returns 2 values, you should capture them both or change method if the second return value is never used. Also use more specific variable name than `return_value`. Here it could be replaced by `templates`.

File: sbc_compassion/tools/patternrecognition.py [hunk 4]
```diff
@@ -167,7 +167,7 @@ def keyPointCenter(keypoints):
         return center / N
 
 
-def find_template(img, templates):
+def find_template(img, templates, threshold=4):
```
Comment (ecino) [inline sbc_compassion/tools/patternrecognition.py]: Why did you add `threshold` if you don't use it inside the function ? Also update the method comment to add description of the parameter if you keep it.

File: sbc_compassion/models/import_letters_history.py [hunk 5]
```diff
@@ -470,8 +480,14 @@ def _find_language(self, img, pattern_center, template):
         if pattern_center is None:
             return
         # get position of the blue corner
-        bluecorner = bcf.BlueCornerFinder(img)
+        box = [float(
+            self.env['ir.config_parameter'].get_param('bluecorner_x_min')),
+               float(
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: ./sbc_compassion/models/import_letters_history.py:485:16: E126 continuation line over-indented for hanging indent

File: sbc_compassion/models/import_letters_history.py (hunk 2 above)
Comment (loikki) [inline sbc_compassion/models/import_letters_history.py]: I had an error with this one, but maybe it was due to something else

**Comment by ecino** [comment]
Tip: don't hesitate to separate commits when your work on completely different tasks. It is not bad to have more than one commits to merge ;)

File: sbc_compassion/models/import_letters_history.py (hunk 2 above)
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: Please remove this and see if it still works (it should)

File: sbc_compassion/data/correspondence_template_data.xml [hunk 6]
```diff
@@ -5,6 +5,43 @@
...
       <field name="layout">CH-A-6S01-1</field>
       <field name="checkbox_ids" eval="False"/>
-    </record>    
+    </record>
+
+    <record id="threshold_keypoints_template" model="ir.config_parameter">
+      <field name="key">threshold_keypoints_template</field>
+      <field name="value">0.25</field>
+    </record>
+
```
Comment (loikki) [inline sbc_compassion/data/correspondence_template_data.xml]: Above, there is the new parameter for the configuration parameters and below are some parameter moved from demo/

File: sbc_compassion/models/correspondence_template.py [hunk 7]
```diff
@@ -107,6 +107,9 @@ class CorrespondenceTemplate(models.Model):
     checkbox_ids = fields.One2many(
         'sponsorship.correspondence.lang.checkbox', 'template_id',
         default=lambda self: self._get_default_checkboxes())
+    nber_keypoints = fields.Integer(
+        "Number of key points", compute="_compute_template_keypoints",
+        store=True)
```
Comment (loikki) [inline sbc_compassion/models/correspondence_template.py]: I added a new parameter for the template in order to know how much keypoints are needed during the detection

File: sbc_compassion/models/correspondence_template.py [hunk 8]
```diff
@@ -267,6 +271,17 @@ def _compute_detection(self):
                             Style.bluesquare_color[2])
                 cv2.circle(img, (self.bluesquare_x, self.bluesquare_y),
                            radius, Style.bluesquare_color)
```
Comment (loikki) [inline sbc_compassion/models/correspondence_template.py]: I am drawing a rectangle for the area of the bluecorner

File: sbc_compassion/models/import_letter_line.py [hunk 9]
```diff
@@ -25,8 +25,6 @@ class ImportLetterLine(models.Model):
 
     sponsorship_id = fields.Many2one('recurring.contract', 'Sponsorship',
                                      compute='_set_sponsorship_id')
```
Comment (loikki) [inline sbc_compassion/models/import_letter_line.py]: Old code written when I did not know well enough odoo.
Now it is more easy (I am just using the bool return by the fields for knowing if they exist)

File: sbc_compassion/models/import_letters_history.py [hunk 10]
```diff
@@ -271,9 +271,15 @@ def _analyze_attachment(self, file_, filename):
         # now try to find the layout
         # loop over all the patterns in the pattern directory
         template, pattern_center = self._find_template(img)
-        lang_id = self._find_language(img, pattern_center, template)
-
-        if self.force_template:
+        if template != self.env.ref('sbc_compassion.default_template'):
+            lang_id = self._find_language(img, pattern_center, template)
```
Comment (loikki) [inline sbc_compassion/models/import_letters_history.py]: Now _find_language returns a list of languages (more easy for travis)

File: sbc_compassion/models/import_letters_history.py [hunk 11]
```diff
@@ -348,15 +354,17 @@ def _find_qrcode(self, img, MULTIPAGE, file_,
         if qrcode is None:
             # rotate image
             img = img[::-1, ::-1]
-            # save it for zxing
-            cv2.imwrite(file_png, img)
```
Comment (loikki) [inline sbc_compassion/models/import_letters_history.py]: If QR code is not detected in the bottom-right corner, it makes no sense to save the rotated image

File: sbc_compassion/models/import_letters_history.py [hunk 12]
```diff
@@ -426,7 +434,16 @@ def _find_template(self, img):
         """
         templates = self.env['sponsorship.correspondence.template'].search(
             [('pattern_image', '!=', False)])
-        return pr.find_template(img, templates)
```
Comment (loikki) [inline sbc_compassion/models/import_letters_history.py]: Now the threshold depends on the template

File: sbc_compassion/tools/checkboxreader.py [hunk 13]
```diff
@@ -191,7 +191,7 @@ def connectedComponents(img, connectivity=8):
     # square mask
     mh = (con.shape[0] - 1) / 2
     mw = mh
-    ret = np.zeros((h, w), np.int8)
+    ret = np.zeros((h, w), np.int16)
```
Comment (loikki) [inline sbc_compassion/tools/checkboxreader.py]: I had some error raised due to the int8 (129 -> -128)

File: sbc_compassion/models/import_letter_line.py [hunk 14]
```diff
@@ -58,12 +56,15 @@ def _check_status(self):
         """
         default_template = self.env.ref('sbc_compassion.default_template')
         for line in self:
-            if not line.sponsorship_found:
-                line.status = "no_sponsorship"
-                if not line.child_partner_found:
+            test_template = not line.template_id or (
```
Comment (ecino) [inline sbc_compassion/models/import_letter_line.py]: Change to:
```python
valid_template = line.template_id and not (
    line.template_id == default_template != line.import_id.force_template)
```
which is more clear

File: sbc_compassion/models/import_letter_line.py [hunk 15]
```diff
@@ -58,12 +56,15 @@ def _check_status(self):
...
+                not line.import_id.force_template and (
+                    line.template_id == default_template))
+            if not line.sponsorship_id:
+                if not (line.child_id and line.partner_id):
                     line.status = "no_child_partner"
-            elif not line.template_id or (line.template_id.id ==
-                                          default_template.id):
+                else:
+                    line.status = "no_sponsorship"
+            elif test_template:
```
Comment (ecino) [inline sbc_compassion/models/import_letter_line.py]: `elif not valid_template`

File: sbc_compassion/models/import_letters_history.py [hunk 16]
```diff
@@ -271,9 +271,15 @@ def _analyze_attachment(self, file_, filename):
         # now try to find the layout
         # loop over all the patterns in the pattern directory
         template, pattern_center = self._find_template(img)
-        lang_id = self._find_language(img, pattern_center, template)
-
-        if self.force_template:
+        if template != self.env.ref('sbc_compassion.default_template'):
+            lang_id = self._find_language(img, pattern_center, template)
+            if len(lang_id) == 1:
+                lang_id = lang_id[0].id
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: If you use only id of language, why not changing method `_find_language` in order to return list of ids instead of list of records ?
Furthermore, you can rename `_find_language` to `_find_languages`.
It would as well be good to rename variable `supporter_languages_id` of `import.letter.line` to `letter_language_id` which makes more sense.

File: sbc_compassion/models/import_letters_history.py [hunk 17]
```diff
@@ -426,7 +434,16 @@ def _find_template(self, img):
         """
         templates = self.env['sponsorship.correspondence.template'].search(
             [('pattern_image', '!=', False)])
-        return pr.find_template(img, templates)
+        threshold = float(
+            self.env.ref('sbc_compassion.threshold_keypoints_template').value)
+        threshold = 0.29
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: Why do you override the threshold defined in the settings ?

File: sbc_compassion/models/import_letters_history.py [hunk 18]
```diff
@@ -426,7 +434,16 @@ def _find_template(self, img):
...
+        return_value = pr.find_template(img, templates, threshold=threshold)
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: would be better to capture result this way:
```
template, pattern_center = pr.find_template(...)
if template is None:
    pattern_center = ...
    template = ...
return template, pattern_center
```

File: sbc_compassion/models/import_letters_history.py (hunk 17 above)
Comment (loikki) [inline sbc_compassion/models/import_letters_history.py]: Thanks I was trying to figure out why my modifications did not work ^^'
It was during the test (faster this way than in xml or gui)

**Comment by ecino** [comment]
On correspondence_template.py, L139, replace `self.nber_keypoints = pattern_keypoints.shape[0]` with `template.nber_keypoints = ...`

**Comment by loikki** [comment]
I have done the modifications, but I think it will be better to find a good value for the threshold (I am currently working on it)

File: sbc_compassion/models/import_letters_history.py [hunk 19]
```diff
@@ -348,15 +354,17 @@ def _find_qrcode(self, img, MULTIPAGE, file_,
...
             qrcode = zx.decode(file_png, try_harder=True, crop=[
                 int(left), int(top), int(width), int(height)])
             # if the QR code is found
             if qrcode is not None:
+                # save it for zxing
+                cv2.imwrite(file_png, img)
                 # replace the image by the returned one
                 data = self._save_img(MULTIPAGE, file_, img)
+            else:
+                qrcode = zx.decode(file_png, try_harder=True)
```
Comment (ecino) [inline sbc_compassion/models/import_letters_history.py]: At line 373, add a condition to the search of partners :
`('is_company', '=', False)`

**Comment by loikki** [comment]
I have done all the modifications and pushed everything (the threshold is at 50%)

File: sbc_compassion/models/import_letter_line.py [hunk 20]
```diff
@@ -25,14 +25,12 @@ class ImportLetterLine(models.Model):
...
                                      compute='_set_sponsorship_id')
-    sponsorship_found = fields.Boolean(compute='_set_sponsorship_id')
-    child_partner_found = fields.Boolean(compute='_set_sponsorship_id')
     partner_id = fields.Many2one('res.partner', 'Partner')
     name = fields.Char(compute='_set_name')
     child_id = fields.Many2one('compassion.child', 'Child')
     template_id = fields.Many2one(
         'sponsorship.correspondence.template', 'Template')
-    supporter_languages_id = fields.Many2one(
+    letter_languages_id = fields.Many2one(
```
Comment (ecino) [inline sbc_compassion/models/import_letter_line.py]: letter_language~~s~~_id, as there is only one language that can be selected.

**Comment by ecino** [comment]
:+1:
